### PR TITLE
GODRIVER-1715 sum up results from batches

### PR DESCRIPTION
**Pull request by iwysiu**
GODRIVER-1715
GODRIVER-1758

### CODE REVIEW

File: mongo/database.go [hunk 1]
```diff
@@ -192,8 +192,9 @@ func (db *Database) RunCommand(ctx context.Context, runCommand interface{}, opts
 	}
 
 	err = op.Execute(ctx)
+	_, convErr := processWriteError(err)
```
Comment (divjotarora) [inline mongo/database.go:195]: Add a comment explaining why we're using processWriteError rather than replaceErrors even though this function doesn't execute a write.

File: mongo/integration/collection_test.go [hunk 2]
```diff
@@ -1154,6 +1154,59 @@ func TestCollection(t *testing.T) {
 				mt.Fatalf("expected BulkWrite error %v, got %v", mongo.ErrUnacknowledgedWrite, err)
 			}
 		})
+		mt.Run("insert and delete with batches", func(mt *mtest.T) {
+			// // TODO(GODRIVER-425): remove this as part a larger project to
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:1158]: Double `//`'s here and below?

File: mongo/integration/collection_test.go [hunk 3]
```diff
@@ -1154,6 +1154,59 @@ func TestCollection(t *testing.T) {
...
+					{"a", int32(i)},
+					{"b", int32(i * 2)},
+					{"c", int32(i * 3)},
+				}
+				insertModels = append(insertModels, mongo.NewInsertOneModel().SetDocument(d))
+				deleteModels = append(deleteModels, mongo.NewDeleteOneModel().SetFilter(bson.D{}))
+			}
+			res, err := mt.Coll.BulkWrite(context.Background(), insertModels)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+			assert.Equal(mt, int64(numDocs), res.InsertedCount, "expected %v inserted documents, got %v", numDocs, res.InsertedCount)
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:1179]: We also need to assert that the `BulkWrite` call sends multiple requests. Otherwise, we don't know if the results are correct because the numbers were merged correctly or because we incorrectly sent them as a single request.

File: mongo/integration/collection_test.go [hunk 4]
```diff
@@ -1154,6 +1154,59 @@ func TestCollection(t *testing.T) {
...
+
+			res, err = mt.Coll.BulkWrite(context.Background(), deleteModels)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+			assert.Equal(mt, int64(numDocs), res.DeletedCount, "expected %v inserted documents, got %v", numDocs, res.DeletedCount)
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:1183]: nit: "inserted" -> "deleted"

File: mongo/integration/collection_test.go [hunk 5]
```diff
@@ -1154,6 +1154,59 @@ func TestCollection(t *testing.T) {
...
+			for i := 0; i < numDocs; i++ {
+				d := bson.D{
+					{"a", int32(i)},
+					{"b", int32(i * 2)},
+					{"c", int32(i * 3)},
+				}
+				insertModels = append(insertModels, mongo.NewInsertOneModel().SetDocument(d))
+				deleteModels = append(deleteModels, mongo.NewDeleteOneModel().SetFilter(bson.D{}))
+			}
+			res, err := mt.Coll.BulkWrite(context.Background(), insertModels)
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:1177]: nit: Use `mtest.Background` here and elsewhere for consistency.

File: mongo/integration/collection_test.go [hunk 6]
```diff
@@ -1154,6 +1154,59 @@ func TestCollection(t *testing.T) {
...
+			// refactor integration and other longrunning tasks.
+			if os.Getenv("EVR_TASK_ID") == "" {
+				mt.Skip("skipping long running integration test outside of evergreen")
+			}
+
+			var models []mongo.WriteModel
+			numModels := 100050
+			// it's significantly faster to upsert one model and modify the rest than to upsert all of them
+			for i := 0; i < numModels; i++ {
+				models = append(models, mongo.NewUpdateOneModel().
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:1196]: nit: This is a pretty long line and it's kind of hard to follow. Let's split up the update document declaration, model construction, and `append` calls into separate lines.

File: mongo/integration/collection_test.go [hunk 7]
```diff
@@ -1154,6 +1154,59 @@ func TestCollection(t *testing.T) {
...
+					SetUpdate(bson.D{{"$set",
+						bson.D{
+							{"a", int32(i + 1)},
+							{"b", int32(i * 2)},
+							{"c", int32(i * 3)},
+						}}}).SetUpsert(true))
+			}
+			res, err := mt.Coll.BulkWrite(context.Background(), models)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+			assert.Equal(mt, int64(1), res.UpsertedCount, "expected %v inserted documents, got %v", 1, res.UpsertedCount)
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:1207]: nit: "inserted" -> "updated"

File: mongo/integration/collection_test.go [hunk 8]
```diff
@@ -1154,6 +1154,59 @@ func TestCollection(t *testing.T) {
...
+			}
+			res, err := mt.Coll.BulkWrite(context.Background(), insertModels)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+			assert.Equal(mt, int64(numDocs), res.InsertedCount, "expected %v inserted documents, got %v", numDocs, res.InsertedCount)
+
+			res, err = mt.Coll.BulkWrite(context.Background(), deleteModels)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+			assert.Equal(mt, int64(numDocs), res.DeletedCount, "expected %v inserted documents, got %v", numDocs, res.DeletedCount)
+		})
+		mt.Run("update with batches", func(mt *mtest.T) {
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:1185]: Should this test also check `MatchedCount`?

File: mongo/integration/collection_test.go (hunk 7 above)
Comment (divjotarora) [inline mongo/integration/collection_test.go:1207]: The upsert is only in the first batch, so we're not testing that `UpsertedCount` is correctly calculated in a multi-batch case. We can either fix this by adding a new model that causes an upsert after all of the existing ones so it'll be in the second batch or by adding unit tests to the `operation` package so we don't have to integration test this.

File: mongo/integration/database_test.go [hunk 9]
```diff
@@ -88,6 +88,35 @@ func TestDatabase(t *testing.T) {
...
+					},
+				},
+			})
+			cmd := bson.D{
+				{"insert", "test"},
+				{"documents", bson.A{bson.D{{"a", 1}}}},
+			}
+			res, gotErr := mt.DB.RunCommand(mtest.Background, cmd).DecodeBytes()
+
+			n, err := res.LookupErr("n")
```
Comment (divjotarora) [inline mongo/integration/database_test.go:110]: nit: The type/error checks can be condensed by using

```
n, ok := res.Lookup("n").Int32OK()
```

File: x/mongo/driver/operation.go [hunk 10]
```diff
@@ -418,6 +415,10 @@ func (op Operation) Execute(ctx context.Context, scratch []byte) error {
 				continue
 			}
 
+			if op.ProcessResponseFn != nil {
```
Comment (divjotarora) [inline x/mongo/driver/operation.go:418]: It's not obvious that the placement of this call matters, so let's add a comment to explain that.

File: mongo/integration/collection_test.go (hunk 3 above)
Comment (iwysiu) [reply mongo/integration/collection_test.go:1181]: done

**Comment by Ankush-Goyal** [comment]
Just curious to know about a scenario where there are partial batch success, how are we handling this.
Scenario: Batch 1 -> Batch 2 -> Batch 3 -> Batch 4
Result: Batch 1(Success) -> Batch 2(Failed) -> Batch 3 (Success) -> Batch 4(Failed).
Failure may be due to connectivity issue etc.
While retrying seems to be there, will it still handle this particular case

File: mongo/integration/collection_test.go [hunk 11]
```diff
@@ -1154,6 +1155,88 @@ func TestCollection(t *testing.T) {
...
+			}
+			mt.ClearEvents()
+			res, err := mt.Coll.BulkWrite(mtest.Background, insertModels)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+			assert.Equal(mt, int64(numDocs), res.InsertedCount, "expected %v inserted documents, got %v", numDocs, res.InsertedCount)
+			mt.FilterStartedEvents(func(evt *event.CommandStartedEvent) bool {
+				return evt.CommandName == "insert"
+			})
+			inserts := len(mt.GetAllStartedEvents())
+			assert.True(mt, inserts > 1, "expected multiple batches, got %v", inserts)
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:1186]: Do we have to assert that it's generically greater than 1? I thought we knew that the number of commands sent would always be 2?

Comment (iwysiu) [reply mongo/integration/collection_test.go]: done

**Comment by iwysiu** [comment]
> Just curious to know about a scenario where there are partial batch success, how are we handling this.
> Scenario: Batch 1 -> Batch 2 -> Batch 3 -> Batch 4
> Result: Batch 1(Success) -> Batch 2(Failed) -> Batch 3 (Success) -> Batch 4(Failed).
> Failure may be due to connectivity issue etc.
> While retrying seems to be there, will it still handle this particular case

Hi @Ankush-Goyal !
For ordered commands, it'll stop after the first batch fails, but will process any successes in that batch, and for unordered commands it'll process the successes in the partially failed batch and then run the next batch. For a given batch, the response is processed unless the batch is being retried, and if a batch fails its last retry the response will be processed then.

File: mongo/integration/collection_test.go (hunk 11 above)
Comment (iwysiu) [reply mongo/integration/collection_test.go:1186]: It turns out that it is not guaranteed to be 2 batches for server versions >= 3.4

File: mongo/integration/collection_test.go [hunk 12]
```diff
@@ -1154,6 +1155,88 @@ func TestCollection(t *testing.T) {
...
+			models = append(models, mongo.NewUpdateOneModel().
+				SetFilter(bson.D{{"x", int32(1)}}).
+				SetUpdate(bson.D{{"$set", bson.D{{"x", int32(1)}}}}).
+				SetUpsert(true),
+			)
+
+			mt.ClearEvents()
+			res, err := mt.Coll.BulkWrite(mtest.Background, models)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+			assert.Equal(mt, int64(2), res.UpsertedCount, "expected %v upserted documents, got %v", 2, res.UpsertedCount)
```
Comment (kevinAlbs) [inline mongo/integration/collection_test.go:1231]: I think I am missing something. I thought the for loop on 1208 matches and updates `numModels-1` without upserting, but the last operation is an upsert. How is this two upserts?

File: mongo/integration/collection_test.go [hunk 13]
```diff
@@ -1154,6 +1155,88 @@ func TestCollection(t *testing.T) {
...
+			// TODO(GODRIVER-425): remove this as part a larger project to
+			// refactor integration and other longrunning tasks.
+			if os.Getenv("EVR_TASK_ID") == "" {
+				mt.Skip("skipping long running integration test outside of evergreen")
+			}
+
+			var models []mongo.WriteModel
+			numModels := 100050
+			// it's significantly faster to upsert one model and modify the rest than to upsert all of them
+			for i := 0; i < numModels-1; i++ {
```
Comment (kevinAlbs) [inline mongo/integration/collection_test.go:1208]: Basic question, but is it possible to run individual tests with the test runner? If "update with batches" can be run in isolation of other tests, it should not rely on "insert and delete with batches" from being run first.

File: mongo/integration/collection_test.go (hunk 12 above)
Comment (kevinAlbs) [reply mongo/integration/collection_test.go:1231]: Thanks for the explanation!

File: mongo/integration/collection_test.go [hunk 14]
```diff
@@ -1154,6 +1155,88 @@ func TestCollection(t *testing.T) {
...
+				SetUpdate(bson.D{{"$set", bson.D{{"x", int32(1)}}}}).
+				SetUpsert(true),
+			)
+
+			mt.ClearEvents()
+			res, err := mt.Coll.BulkWrite(mtest.Background, models)
+			assert.Nil(mt, err, "BulkWrite error: %v", err)
+			assert.Equal(mt, int64(2), res.UpsertedCount, "expected %v upserted documents, got %v", 2, res.UpsertedCount)
+			assert.Equal(mt, int64(numModels-2), res.ModifiedCount, "expected %v modified documents, got %v", numModels-2, res.ModifiedCount)
+			assert.Equal(mt, int64(numModels-2), res.MatchedCount, "expected %v matched documents, got %v", numModels-2, res.ModifiedCount)
```
Comment (kevinAlbs) [inline mongo/integration/collection_test.go:1233]: Given that UpsertedCount was incorrect for merging multiple batches, I imagine we do not have coverage on all BulkWriteResult fields when batch splitting occurs. Can we check that `res.UpsertedIDs` has two entries as well, and use those entries to retrieve the expected upserted documents?

**Review by kevinAlbs — APPROVED** [review]
LGTM with an additional test assertion.

File: mongo/integration/collection_test.go (hunk 11 above)
Comment (divjotarora) [inline mongo/integration/collection_test.go:1186]: Add a comment to describe why we're making a generic assertion rather than asserting that `inserts` is 2. I think this will be confusing for future readers.

**Review by divjotarora — APPROVED** [review]
LGTM mod the one small comment I have open.

**Comment by iwysiu** [comment]
@divjotarora @kevinAlbs the last commit ended up being more than test changes, so maybe another quick look?

File: mongo/integration/collection_test.go [hunk 15]
```diff
@@ -1154,6 +1155,124 @@ func TestCollection(t *testing.T) {
...
+			assert.Equal(mt, int64(2), res.UpsertedCount, "expected %v upserted documents, got %v", 2, res.UpsertedCount)
+			assert.Equal(mt, 2, len(res.UpsertedIDs), "expected %v upserted ids, got %v", 2, len(res.UpsertedIDs))
+
+			id1, ok := res.UpsertedIDs[0]
+			assert.True(mt, ok, "expected id at key 0")
+			id2, ok := res.UpsertedIDs[int64(numModels-1)]
+			assert.True(mt, ok, "expected id at key %v", numModels-1)
+
+			doc, err := mt.Coll.FindOne(mtest.Background, bson.D{{"_id", id1}}).DecodeBytes()
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:1252]: Two ideas to clean this up a bit. I don't have a preference between them:

1. Use the `Int32OK` variants to avoid doing a separate `LookupErr` call: `a, ok := doc.Lookup("a").Int32OK()`.

2. Use `compareDocs` instead:

```
expected := {"a": int32(numModels-1), "b": int32((numModels-2)*2), "c": int32((numModels-2)*3)}
compareDocs(expected, doc)
```

You'd have to build `expected` either using `bson.D` followed by `bson.Marshal` or using `bsoncore` directly (but not the `DocumentBuilder` type because that can't be backported to the 1.4 branch).

Comment (divjotarora) [inline mongo/integration/collection_test.go:1252]: Leave a comment above this saying that we're finding the upserted documents and making sure they are what we expect.

File: x/mongo/driver/batch_cursor.go [hunk 16]
```diff
@@ -266,7 +266,7 @@ func (bc *BatchCursor) getMore(ctx context.Context) {
 		},
 		Database:   bc.database,
 		Deployment: SingleServerDeployment{Server: bc.server},
-		ProcessResponseFn: func(response bsoncore.Document, srvr Server, desc description.Server) error {
+		ProcessResponseFn: func(response bsoncore.Document, srvr Server, desc description.Server, currIndex int) error {
```
Comment (divjotarora) [inline x/mongo/driver/batch_cursor.go:269]: Here and in every other operation file besides `update.go`: this field is ignored so we should express that by marking it as `_` in the function signature rather than giving it a name

File: mongo/integration/collection_test.go [hunk 17]
```diff
@@ -1154,6 +1155,124 @@ func TestCollection(t *testing.T) {
...
+				return evt.CommandName == "delete"
+			})
+			// MaxWriteBatchSize changed between 3.4 and 3.6, so there isn't a given number of batches that this will be split into
+			deletes := len(mt.GetAllStartedEvents())
+			assert.True(mt, deletes > 1, "expected multiple batches, got %v", deletes)
+		})
+		mt.Run("update with batches", func(mt *mtest.T) {
+			// TODO(GODRIVER-425): remove this as part a larger project to
+			// refactor integration and other longrunning tasks.
+			// if os.Getenv("EVR_TASK_ID") == "" {
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply mongo/integration/collection_test.go]: done

File: mongo/integration/collection_test.go (hunk 15 above)
Comment (iwysiu) [reply mongo/integration/collection_test.go:1241]: done

Comment (iwysiu) [reply mongo/integration/collection_test.go:1241]: done

File: x/mongo/driver/batch_cursor.go (hunk 16 above)
Comment (iwysiu) [reply x/mongo/driver/batch_cursor.go:269]: done

**Review by divjotarora — APPROVED** [review]
